### DISCUSSION TITLE
Dynamically render input boxes based on fetched API results without useState?

### DISCUSSION THREAD

**Discussion by risafj** [react-hook-form/react-hook-form#2104] [Q&A] [answered]
Hi @bluebill1049 and everyone, thanks for this great library 😄 

I have a question about registering checkboxes dynamically based on an API call.
The [Sandbox](https://codesandbox.io/s/react-hook-form-demo-p6oth?file=/src/App.js) is self-explanatory, but basically what I'm trying to do is:

1. Call API for data (like below) necessary for rendering a list of checkboxes
```js
[
  { id: 1, name: "cats", label: "Cats", value: true },
  { id: 2, name: "dogs", label: "Dogs", value: false },
  { id: 3, name: "parrots", label: "Parrots", value: false }
];
```

2. Render some checkboxes based on the fetched data
```jsx
  const petCheckboxes = petsData.map(petData => (
    <li key={petData.id}>
      <input
        type="checkbox"
        defaultChecked={petData.value}
        name={petData.name}
        ref={register}
        id={petData.id}
      />
      <label htmlFor={petData.id}>{petData.label}</label>
    </li>
  ));
```

Right now, I'm storing the result of the API call in `const [petsData, setPetsData] = useState([]);`. But this implementation seems messy to me because I don't actually need `petsData` stored - I'm only using it to trigger the `petsData.map()`. React Hook Form is the one that keeps track of the state of the checkboxes.
My question is, is there a better way to do this so there's need for `useState`?

**Comment by bluebill1049** [comment, marked as answer]
I think you will need useState with this because you need a refresh when API resolves.

**Comment by risafj** [comment]
@bluebill1049 Got it. Thank you for your quick response!
